verifier/sources/gps/gps_statement_verifier.move

```diff
@@ -0,0 +1,476 @@
+                EINVALID_HASH_FOR_MEMORY_PAGE_0
+            );
+            assert!(
+                *borrow(&public_memory_pages, n_pages * PAGE_INFO_SIZE) == prod,
+                EINVALID_CUMULATIVE_PRODUCT
+            );
+            *checkpoint = CHECKPOINT2_VPAR;
+            return
+        };
+
+        // NOLINTNEXTLINE: reentrancy-benign.
+        if (*checkpoint == CHECKPOINT2_VPAR) {
+            let new_cairo_public_input_length = cairo_aux_input_length - 2;
+            let cairo_public_input = *cairo_aux_input;
+            trim_only(&mut cairo_public_input, new_cairo_public_input_length); // z and alpha.
+
+            if (verify_proof_external(signer, proof_params, proof, &cairo_public_input)) {
+                *checkpoint = CHECKPOINT3_VPAR;
+            };
+            return
+        };
+
+        if (*checkpoint == CHECKPOINT3_VPAR) {
+            let new_cairo_public_input_length = cairo_aux_input_length - 2;
+            let cairo_public_input = *cairo_aux_input;
+            trim_only(&mut cairo_public_input, new_cairo_public_input_length); // z and alpha.
+            let public_memory_pages = cairo_public_input;
+            let (public_memory_offset, _) = get_layout_info();
+            trim_head(&mut public_memory_pages, (public_memory_offset as u64));
+
+            register_gps_facts(
+                signer,
+                task_metadata,
+                &public_memory_pages,
+                *borrow(cairo_aux_input, OFFSET_OUTPUT_BEGIN_ADDR)
+            );
+            emit(VparFinished {
+                ok: true
+            });
+            *checkpoint = CHECKPOINT1_VPAR;
+        };
+    }
+
+    // Registers the fact for memory page 0, which includes:
+    // 1. The bootloader program,
+    // 2. Arguments and return values of main()
+    // 3. Some of the data required for computing the task facts. which is represented in
+    //    taskMetadata.
+    // Returns information on the registered fact.
+    //
+    // Arguments:
+    //   selectedBuiltins: A bit-map of builtins that are present in the layout.
+    //       See CairoVerifierContract.sol for more information.
+    //   taskMetadata: Per task metadata.
+    //   cairoAuxInput: Auxiliary input for the cairo verifier.
+    //
+    // Assumptions: cairoAuxInput is connected to the public input, which is verified by
+    // cairoVerifierContractAddresses.
+    // Guarantees: taskMetadata is consistent with the public memory, with some sanity checks.
+    fun register_public_memory_main_page(
+        signer: &signer,
+        task_metadata: &vector<u256>,
+        cairo_aux_input: &vector<u256>,
+        selected_builtins: u256
+    ): (u256, u256, u256) acquires ConstructorConfig {
+        let n_tasks = *borrow(task_metadata, 0);
+        // Ensure 'n_tasks' is bounded as a sanity check (the bound is somewhat arbitrary).
+        assert!(n_tasks < (1 << 30), EINVALID_NUMBER_OF_TASKS);
+
```

Comment:
fun register_public_memory_main_page(
        signer: &signer,
        task_metadata: &vector<u256>,
        cairo_aux_input: &vector<u256>,
        selected_builtins: u256
    ): (u256, u256, u256) acquires ConstructorConfig {
        let n_tasks = *borrow(task_metadata, 0);
        // Ensure 'n_tasks' is bounded as a sanity check (the bound is somewhat arbitrary).
        assert!(n_tasks < (1 << 30), EINVALID_NUMBER_OF_TASKS);

        // Public memory length.
        let public_memory_length = (PROGRAM_SIZE +
            // return fp and pc =
            2 +
            N_MAIN_ARGS +
            N_MAIN_RETURN_VALUES +
            // Bootloader config size =
            2 +
            // Number of tasks cell =
            1 +
            2 *
                n_tasks);

        let public_memory = vector[];
        // Write public memory, which is a list of pairs (address, value).
        {
            let bootloader_program = get_compiled_program(signer);
            let n = length(&bootloader_program);
            for (i in 0..n) {
                push_back(&mut public_memory, (i + INITIAL_PC as u256));
                push_back(&mut public_memory, *borrow(&bootloader_program, i));
            }
        };

        {
            // Execution segment - Make sure [initial_fp - 2] = initial_fp and .
            // This is required for the safe call feature (that is, all call instructions will
            // return, even if the called function is malicious).
            // It guarantees that it's not possible to create a cycle in the call stack.
            let initial_fp = *borrow(cairo_aux_input, OFFSET_EXECUTION_BEGIN_ADDR);
            assert!(initial_fp >= 2, EINVALID_EXECUTION_BEGIN_ADDRESS);
            push_back(&mut public_memory, initial_fp - 2);
            push_back(&mut public_memory, initial_fp);
            push_back(&mut public_memory, initial_fp - 1);
            push_back(&mut public_memory, 0);

            // Execution segment: Enforce main's arguments and return values.
            // Note that the page hash depends on the order of the (address, value) pair in the
            // public_memory and consequently the arguments must be written before the return values.
            let return_values_address = *borrow(cairo_aux_input, OFFSET_EXECUTION_STOP_PTR) - N_BUILTINS;
            let builtin_segment_info_offset = OFFSET_OUTPUT_BEGIN_ADDR;
            let return_values = vector[];
            for (i in 0..N_BUILTINS) {
                // Write argument address.
                push_back(&mut public_memory, initial_fp + i);

                // Write return value address.
                push_back(&mut return_values, return_values_address + i);

                // Write values.
                if ((selected_builtins & 1) != 0) {
                    // Set the argument to the builtin start pointer.
                    push_back(&mut public_memory, *borrow(cairo_aux_input, builtin_segment_info_offset));
                    push_back(&mut return_values, *borrow(cairo_aux_input, builtin_segment_info_offset + 1));
                    // Set the return value to the builtin stop pointer.
                    builtin_segment_info_offset = builtin_segment_info_offset + 2;
                } else {
                    // Builtin is not present in layout, set the argument value and return value to 0.
                    push_back(&mut public_memory, 0);
                    push_back(&mut return_values, 0);
                };
                selected_builtins = selected_builtins >> 1;
            };
            assert!(selected_builtins == 0, ESELECTED_BUILTINS_VECTOR_IS_TOO_LONG);
            append(&mut public_memory, return_values);
            // Skip the return values which were already written.
        };

        // Program output.
        {
            let ConstructorConfig {
                hashed_supported_cairo_verifiers,
                simple_bootloader_program_hash
            } = borrow_global<ConstructorConfig>(address_of(signer));
            let output_address = *borrow(cairo_aux_input, OFFSET_OUTPUT_BEGIN_ADDR);
            // Force that memory[outputAddress] and memory[outputAddress + 1] contain the
            // bootloader config (which is 2 words size).
            push_back(&mut public_memory, output_address);
            push_back(&mut public_memory, *simple_bootloader_program_hash);
            push_back(&mut public_memory, output_address + 1);
            push_back(&mut public_memory, *hashed_supported_cairo_verifiers);
            push_back(&mut public_memory, output_address + 2);
            push_back(&mut public_memory, n_tasks);
            output_address = output_address + 3;

            let current_metadata_offset = METADATA_TASKS_OFFSET;

            for (task in 0..n_tasks) {
                let output_size = *borrow(
                    task_metadata,
                    current_metadata_offset + METADATA_OFFSET_TASK_OUTPUT_SIZE
                );

                // Ensure 'outputSize' is at least 2 and bounded from above as a sanity check
                // (the bound is somewhat arbitrary).
                assert!(2 <= output_size && output_size < (1 << 30), EINVALID_TASK_OUTPUT_SIZE);
                let program_hash = *borrow(
                    task_metadata,
                    current_metadata_offset + METADATA_OFFSET_TASK_PROGRAM_HASH
                );
                let n_tree_pairs = *borrow(
                    task_metadata,
                    current_metadata_offset + METADATA_OFFSET_TASK_N_TREE_PAIRS
                );

                // Ensure 'nTreePairs' is at least 1 and bounded from above as a sanity check
                // (the bound is somewhat arbitrary).
                assert!(
                    1 <= n_tree_pairs && n_tree_pairs < (1 << 20),
                    EINVALID_NUMBER_OF_PAIRS_IN_MERKLE_TREE_STRUCTURE
                );
                // Force that memory[outputAddress] = outputSize.
                push_back(&mut public_memory, output_address);
                push_back(&mut public_memory, output_size);
                push_back(&mut public_memory, output_address + 1);
                push_back(&mut public_memory, program_hash);

                output_address = output_address + output_size;

                current_metadata_offset = METADATA_TASK_HEADER_SIZE + (2 * n_tree_pairs as u64) + current_metadata_offset;
            };

            assert!(length(task_metadata) == current_metadata_offset, EINVALID_LENGTH_OF_TASK_METADATA);

            assert!(
                *borrow(cairo_aux_input, OFFSET_OUTPUT_STOP_PTR) == output_address,
                EINCONSISTENT_PROGRAM_OUTPUT_LENGTH
            );
        };
        assert!(
            (length(&public_memory) as u256) == MEMORY_PAIR_SIZE * public_memory_length,
            ENOT_ALL_CAIRO_PUBLIC_INPUTS_WERE_WRITTEN
        );
        let cairo_aux_input_length = length(cairo_aux_input);
        let z = *borrow(cairo_aux_input, cairo_aux_input_length - 2);
        let alpha = *borrow(cairo_aux_input, cairo_aux_input_length - 1);
        let (memory_hash, prod) = register_regular_memory_page(
            signer,
            &public_memory,
            z,
            alpha
        );
        return (public_memory_length, memory_hash, prod)
    }